File: dotnet/Multiplayer/Rooms/Actions/ActionSubmitRequest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace Michitai.Multiplayer.Rooms.Actions
{
    /// <summary>
    /// Request data for submitting an action to players in a room.
    /// Serializes action data for API transmission.
    /// </summary>
    /// <typeparam name="T">The type of request data.</typeparam>
    public class ActionSubmitRequest<T> where T : class, new()
    {
        [JsonInclude]
        private string Target_players { get; set; } = ERoomTargetPlayers.Host.ToString().ToLower();
        [JsonInclude]
        private int[]? Target_players_ids { get; set; }
        [JsonInclude]
        private string Action_type { get; set; } = string.Empty;
        [JsonInclude]
        private T? Request_data { get; set; }

        /// <summary>
        /// Initializes a new ActionSubmitRequest.
        /// </summary>
        /// <param name="targetPlayers">Which players to target with the action.</param>
        /// <param name="type">The type of action being submitted.</param>
        /// <param name="data">The request data for the action.</param>
        /// <param name="targetPlayersIds">Specific player IDs to target with the action.</param>
        public ActionSubmitRequest(ERoomTargetPlayers targetPlayers, string type, T? data = null, int[]? targetPlayersIds = null)
        {
            this.Target_players = targetPlayers.ToString().ToLower();
            this.Target_players_ids = targetPlayersIds;
            this.Action_type = type;
            this.Request_data = data;
        }
    }
}

File: java/src/main/java/com/michitai/multiplayer/rooms/actions/ActionSubmitRequest.java
package com.michitai.multiplayer.rooms.actions;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Request for submitting an action to target players.
 */
public class ActionSubmitRequest {
    @JsonProperty("target_players")
    private String targetPlayers;

    @JsonProperty("action_type")
    private String actionType;

    @JsonProperty("request_data")
    private String requestData;

    @JsonProperty("target_players_ids")
    private int[] targetPlayersIds;

    public ActionSubmitRequest(ERoomTargetPlayers targetPlayers, String actionType, String requestData, int[] targetPlayersIds) {
        this.targetPlayers = targetPlayers.name().toLowerCase();
        this.actionType = actionType;
        this.requestData = requestData;
        this.targetPlayersIds = targetPlayersIds;
    }

    public String getTargetPlayers() {
        return targetPlayers;
    }

    public void setTargetPlayers(String targetPlayers) {
        this.targetPlayers = targetPlayers;
    }

    public String getActionType() {
        return actionType;
    }

    public void setActionType(String actionType) {
        this.actionType = actionType;
    }

    public String getRequestData() {
        return requestData;
    }

    public void setRequestData(String requestData) {
        this.requestData = requestData;
    }

    public int[] getTargetPlayersIds() {
        return targetPlayersIds;
    }

    public void setTargetPlayersIds(int[] targetPlayersIds) {
        this.targetPlayersIds = targetPlayersIds;
    }
}
